services/s3/src/test/java/software/amazon/awssdk/services/s3/internal/crossregion/S3CrossRegionSyncClientRedirectTest.java

```diff
@@ -0,0 +1,269 @@
+    }
+
+    @Test
+    void requestsWithNoBucketsDoesnotOverrideEndpoints() {
+        when(mockDelegateClient.serviceClientConfiguration()).thenReturn(ENDPOINT_CONFIGURED);
+
+        when(mockDelegateClient.listBuckets(any(ListBucketsRequest.class)))
+            .thenThrow(redirectException(301, CROSS_REGION));
+
+        assertThatExceptionOfType(S3Exception.class)
+            .isThrownBy(() -> decoratedS3Client.listBuckets(ListBucketsRequest.builder().build()))
+            .withMessageContaining("Status Code: 301");
+
+        ArgumentCaptor<ListBucketsRequest> requestArgumentCaptor = ArgumentCaptor.forClass(ListBucketsRequest.class);
+        verify(mockDelegateClient, times(1)).listBuckets(requestArgumentCaptor.capture());
+        assertThat(requestArgumentCaptor.getAllValues().get(0).overrideConfiguration()).isNotPresent();
+    }
+
+    @Test
+    void bucketRegionChanged_updatesCache_AndDoesNotError() throws ExecutionException, InterruptedException {
+        when(mockDelegateClient.serviceClientConfiguration()).thenReturn(ENDPOINT_CONFIGURED);
+        when(mockDelegateClient.listObjects(any(ListObjectsRequest.class)))
+            .thenThrow(redirectException(301, CROSS_REGION))
+            .thenReturn(ListObjectsResponse.builder().contents(S3_OBJECTS).build())
+            .thenThrow(redirectException(301, CHANGED_CROSS_REGION))
+            .thenReturn(ListObjectsResponse.builder().contents(S3_OBJECTS).build());
+
+        // first call to update the local cache
+        ListObjectsResponse preCacheCall = decoratedS3Client.listObjects(i -> i.bucket(CROSS_REGION_BUCKET));
+        assertThat(preCacheCall.contents()).isEqualTo(S3_OBJECTS);
+
+        ArgumentCaptor<ListObjectsRequest> requestArgumentCaptor = ArgumentCaptor.forClass(ListObjectsRequest.class);
+        verify(mockDelegateClient, times(2)).listObjects(requestArgumentCaptor.capture());
+        // Verification for first call
+        assertThat(requestArgumentCaptor.getAllValues().get(0).overrideConfiguration()).isNotPresent();
+
+        // Verification of fall back for  first call
+        EndpointProvider overriddenEndpointProviderPreCache =
+            requestArgumentCaptor.getAllValues().get(1).overrideConfiguration().get().endpointProvider().get();
+        assertThat(overriddenEndpointProviderPreCache).isInstanceOf(BucketEndpointProvider.class);
+
+        assertThat(((S3EndpointProvider) overriddenEndpointProviderPreCache)
+                       .resolveEndpoint(e -> e.region(Region.US_WEST_2).bucket(CROSS_REGION_BUCKET).build()).get().url().getHost())
+            .isEqualTo("s3." + CROSS_REGION + ".amazonaws.com");
+        // Second call to use the local cache
+        ListObjectsResponse afterRegionUpdate = decoratedS3Client.listObjects(i -> i.bucket(CROSS_REGION_BUCKET));
+        assertThat(afterRegionUpdate.contents()).isEqualTo(S3_OBJECTS);
+        requestArgumentCaptor = ArgumentCaptor.forClass(ListObjectsRequest.class);
```

Comment:
Do we need to initialize the captor twice?

Reply:
removed